src/main/java/frc/robot/subsystems/turret/Turret.java

```diff
@@ -0,0 +1,161 @@
+   *     the turret position from two encoders
+   * @param cancoder1
+   * @param cancoder2
+   * @return rotation2d of the turret position
+   */
+  public static Rotation2d getTurretPosition(Rotation2d cancoder1, Rotation2d cancoder2) {
+    Rotation2d diff = cancoder2.minus(cancoder1);
+    double drivingRotRaw = diff.getDegrees() / TurretConstants.differenceDegrees;
+    double expectedG1 = (drivingRotRaw * TurretConstants.ratio1) % 1;
+    double g1Diff = expectedG1 - (cancoder1.getRotations() % 1);
+    if (g1Diff > 0.5) {
+      g1Diff -= 1;
+    } else if (g1Diff < -0.5) {
+      g1Diff += 1;
+    }
+    double drivingRem = g1Diff / TurretConstants.ratio1;
+    return Rotation2d.fromRotations(drivingRotRaw - drivingRem);
+  }
+
+  /**
+   * @param position Desired position of the turret motor
+   * @param swerveAngle Current angle of the swerve drivetrain
+   * @return command to set the position of the turret absolute to the swerve angle
+   */
+  public Command setPositionRelativeToSwerve(Rotation2d position, CommandSwerveDrivetrain swerve) {
+    return this.run(
+        () ->
+            turretIO.setPosition(
+                position.minus(swerve.getState().Pose.getRotation()).getRotations()
+                    / TurretConstants.gearRatio));
+  }
+
+  /**
+   * @param position Desired position of the turret motor
+   * @return Command to set the turret motor to the desired position / by the gear ratio
+   */
+  public Command setPosition(Rotation2d position) {
+    return this.runOnce(
+        () -> turretIO.setPosition(position.getRotations() / TurretConstants.gearRatio));
+  }
+
+  /**
+   * @return Command to zero the turret motor's encoder position
+   */
+  public Command zero() {
+    return this.runOnce(() -> turretIO.zero());
+  }
+
+  /**
+   * Will run a PID loop that tracks the tx of the speaker tag and attempts to get it to 0 by moving
+   * the turret so the LL crosshair is centered
+   *
+   * @param vision Reference to the vision subsystem
+   * @return Command to run a PID loop to lock the turret to the speaker tag
+   */
+  public Command lockToSpeakerTag(Vision vision) {
+    return new PIDCommand(
+        new PIDController(
+            TurretConstants.followTagP, TurretConstants.followTagI, TurretConstants.followTagD),
+        vision::getCompensatedCenterLimelightX,
+        0,
+        output -> turretIO.setVoltage(output),
+        this);
+  }
+
+  /**
+   * @param swerve Reference to the swerve drivetrain
+   * @param speakerPose Pose of the speaker
+   * @return Command to turn turret to face the speaker
+   */
+  public Command lockToSpeakerPose(CommandSwerveDrivetrain swerve, Pose2d speakerPose) {
+    return this.run(
+        () ->
+            this.setPosition(
+                swerve
+                    .getState()
+                    .Pose
+                    .rotateBy(
+                        getTurretPosition(
```

Comment:
you need to 0 the turret at a specific direction in order for this to work i believe - (or offsets?)